File: TypeScript/src/session2.td/heritageMultiple/AlgebreNatRecursif.ts
import {Nat} from "../../session1.td/Nat";
import {AlgebreNatRecursifSuccesseur} from "./AlgebreNatRecursifSuccesseur";
import {AlgebreNatRecursifZero} from "./AlgebreNatRecursifZero";

export interface AlgebreNatRecursif extends AlgebreNatRecursifSuccesseur, AlgebreNatRecursifZero{
    somme(n : Nat) : Nat;
    zero() : Nat;
    produit(n : Nat) : Nat;
    un() : Nat;
    modulo(n : Nat) : Nat;
    div(n : Nat) : Nat;
    equals(o : Nat) : boolean;
    toString() : string;
}

export abstract class AlgebreNatRecursif_default implements AlgebreNatRecursif {
    div(n: Nat): Nat {
        let r : Nat = this.predecesseur().modulo(n);
        let q : Nat = this.predecesseur().div(n);
        return this.creerSuccesseur(r).equals(n) ? this.creerSuccesseur(q) : q;
    }

    equals(o: Nat): boolean {
        return this.val() == o.val();
    }

    modulo(n: Nat): Nat {
        let r : Nat = this.predecesseur().modulo(n);
        return this.creerSuccesseur(r).equals(n) ? this.creerZero() : this.creerSuccesseur(r);
    }

    produit(n: Nat): Nat {
        if (n.estNul() || this.estNul())
            return this.creerZero();
        else
            return n.somme(this.predecesseur().produit(n));
    }

    somme(n: Nat): Nat {
        if (this.equals(this.zero())) {
            return n;
        }
        return this.creerSuccesseur(this.predecesseur().somme(n));
    }

    un(): Nat {
        return this.creerNatAvecRepresentation("1");
    }

    zero(): Nat {
        return this.creerNatAvecRepresentation("0");
    }

    abstract chiffre(i: number): number;

    abstract creerNatAvecRepresentation(repDecimale: String): Nat;

    abstract creerNatAvecValeur(val: number): Nat;

    abstract creerSuccesseur(predecesseur: Nat): Nat;

    abstract creerZero(): Nat;

    abstract estNul(): boolean;

    abstract predecesseur(): Nat;

    abstract taille(): number;

    abstract val(): number;

}

File: Java/src/session2/td/heritageMultiple/AlgebreNatRecursif.java
package session2.td.heritageMultiple;
import session1.td.Nat;

public interface AlgebreNatRecursif extends AlgebreNatRecursifSuccesseur,AlgebreNatRecursifZero{

    default String representer() {
        return "S^" + this.val() + "(0)";
    }


    default Nat somme(Nat x) {
        if (this.equals(this.zero())) {
            return x;
        }
        return this.creerSuccesseur(this.predecesseur().somme(x));
    }

    default  Nat produit(Nat x) {
        if (x.estNul() || this.estNul())
            return this.creerZero();
        else
            return x.somme(this.predecesseur().produit(x));
    }

    default Nat modulo(Nat x) {
        if (this.estNul())
            return this.zero();
        else {
            Nat r = this.predecesseur().modulo(x);
            return this.creerSuccesseur(r).equals(x) ? this.creerZero() : this.creerSuccesseur(r);
        }
    }

    default  Nat div(Nat x) {
        if (this.estNul())
            return this.zero();
        else {
            Nat r = this.predecesseur().modulo(x);
            Nat q = this.predecesseur().div(x);
            return this.creerSuccesseur(r).equals(x) ? this.creerSuccesseur(q) : q;
        }
    }
    default boolean estEgal(Object o ){
        if(!(o instanceof Nat))
            return false;
        Nat x = (Nat)o;
        return this.val() == x.val();
    }
    default Nat un() {
        return this.creerSuccesseur(this.zero());
    }

    default Nat zero() {
        return this.creerZero();
    }

}
